packages/aws-cdk/lib/api/logs/logs-monitor.ts

```diff
@@ -0,0 +1,209 @@
+   * multiple environments (account+region), we need to store a list of log groups
+   * per env along with the SDK object that has access to read from
+   * that environment.
+   */
+  public addLogGroups(env: cxapi.Environment, sdk: ISDK, logGroupNames: string[]): void {
+    const awsEnv = `${env.account}:${env.region}`;
+    const logGroupsStartTimes = logGroupNames.reduce((acc, groupName) => {
+      acc[groupName] = this.startTime;
+      return acc;
+    }, {} as { [logGroupName: string]: number });
+    this.envsLogGroupsAccessSettings.set(awsEnv, {
+      sdk,
+      logGroupsStartTimes: {
+        ...this.envsLogGroupsAccessSettings.get(awsEnv)?.logGroupsStartTimes,
+        ...logGroupsStartTimes,
+      },
+    });
+  }
+
+  private scheduleNextTick(sleep: number): void {
+    setTimeout(() => void(this.tick()), sleep);
+  }
+
+  private async tick(): Promise<void> {
+    if (!this.active) {
+      return;
+    }
+    try {
+      const events = flatten(await this.readNewEvents());
+      events.forEach(event => {
+        this.print(event);
+      });
+    } catch (e) {
+      error('Error occurred while monitoring logs: %s', e);
+    }
+    this.scheduleNextTick(SLEEP);
+  }
+
+  /**
+   * Reads all new log events from a set of CloudWatch Log Groups
+   * in parallel
+   */
+  private async readNewEvents(): Promise<Array<Array<CloudWatchLogEvent>>> {
+    const promises: Array<Promise<Array<CloudWatchLogEvent>>> = [];
+    for (const settings of this.envsLogGroupsAccessSettings.values()) {
+      for (const group of Object.keys(settings.logGroupsStartTimes)) {
+        promises.push(this.readEventsFromLogGroup(settings, group));
+      }
+    }
+    return Promise.all(promises);
+  }
+
+  /**
+   * Print out a cloudwatch event
+   */
+  private print(event: CloudWatchLogEvent): void {
+    print(util.format('[%s] %s %s',
+      colors.blue(event.logGroupName),
+      colors.yellow(event.timestamp.toLocaleTimeString()),
+      event.message.trim()));
+  }
+
+  /**
+   * Reads all new log events from a CloudWatch Log Group
+   * starting at either the time the hotswap was triggered or
+   * when the last event was read on the previous tick
+   */
+  private async readEventsFromLogGroup(
+    logGroupsAccessSettings: LogGroupsAccessSettings,
+    logGroupName: string,
+  ): Promise<Array<CloudWatchLogEvent>> {
+    const events: CloudWatchLogEvent[] = [];
+
+    // log events from some service are ingested faster than others
+    // so we need to track the start/end time for each log group individually
+    // to make sure that we process all events from each log group
+    const startTime = logGroupsAccessSettings.logGroupsStartTimes[logGroupName] ?? this.startTime;
+    let endTime = startTime;
+    try {
+      let nextToken: string | undefined;
+      do {
+        const response = await logGroupsAccessSettings.sdk.cloudWatchLogs().filterLogEvents({
+          logGroupName: logGroupName,
+          nextToken,
```

Comment:
the `filterLogEvents` operation returns as many log events after the `startTime` as can fit in 1 MB (up to 100 events since we specified that limit). If the first 90 events took up the 1 MB then we would receive a `nextToken` to retrieve the final 10. In the case where we receive the `nextToken` we still want to use the same `startTime`, we only want to update the `startTime` for the next time we process events (the next time we `tick()`.

Reply:
Interesting (and odd). But okay!

Reply:
Ended up removing the `nextToken` since the intent is to just provide a sampling of 100 events if we have > 100.